doc/sphinx/practical-tools/utilities.rst

```diff
@@ -7,32 +7,307 @@
+remain consistent.
+
+Note that packages marked `released` in the package list web page are more stable
+than those marked `extra-dev`.  To install `extra-dev` packages,
+first add the `coq-extra-dev` opam repository to your local opam installation
+with this command::
+
+  opam repo add coq-extra-dev https://coq.inria.fr/opam/extra-dev
 
-   % ocamlfind ocamlopt -thread -rectypes -linkall -linkpkg \
-                 -package coq.toplevel \
-                 topbin/coqtop_bin.ml -o my_toplevel.native
+While this is the easiest way to install packages, it is not the only way.
+
+You can see the list of installed Coq packages with `opam list|grep coq-`, which
+will show you lines like this containing the package name, its version and short
+description::
+
+   $ opam list | grep coq-
+   coq-bignums               8.15.0          Bignums, the Coq library of arbitrary large numbers
+
+However, you will need to find the :term:`logical name`: used to refer to the package
+in :cmd:`Require` commands.  There are a couple ways to do this:
+
+- If you installed with opam, use :n:`opam show --list-files coq-bignums | head -n1` -
+  the last component of the filename is the logical name (`Bignums`).
+
+- On Linux, :n:`ls $(coqtop -where)/user-contrib` shows the logical names of all
+  installed user-contributed packages.  You should be able to guess which one you
+  need.
+
+- Use the :cmd:`Print LoadPath` command when running Coq, which shows the mapping
+  from :term:`logical path`\s to directories.  Again, you should be able to guess.
+
+The last two methods work even if you didn't install with opam.  Perhaps in the
+future the package name to logical name mapping will be more readily available.
+
+Once you know the logical name of the package, use it to load compiled
+files from the package with the :cmd:`Require` command.
 
 
-For example, to statically link |Ltac|, you can just do:
+Setup for working on your own projects
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The working and master copies of source code for your own projects should
+be kept separate from the Coq installation.  In particular, when you upgrade
+to a new version of Coq, any directories you created in the old version won't
+be copied or moved.
+
+We encourage you to use a source code control system for any non-trivial
+project because it makes it easy to track the history of your changes.
+`git <https://git-scm.com/>`_ is the package most used by Coq projects.
+Typically, each package has its own git repository.
 
-::
+**Single directory projects.**  If your project has a single file (and
+depends only on packages in the Coq installation), put it in any directory
+you like and run Coq from that directory.  If your project has multiple files,
+then you need to compile them in an order that matches their dependencies.
+If B.v depends on A.v, then you should compile A.v before B.v.  You can do this
+with :n:`coqc A.v B.v`, but you may find it tedious to identify the dependencies.
```

Comment:
COQPATH is not really needed by anybody. It simply allows for people to install libraries in places other than user-contrib. If you have a coq_makefile project or dune project, they will both install into user-contrib today by default.

The `Cd` commands in Coq are really ready to be deprecated any day. They unfortunately have one valid use case with extraction today, but this is really a lack of a feature in extraction, than a valid use of Cd. We should refrain as much as possible from mentioning them.

Even a multi-directory project, will not require setting COQPATH. The only reason you might want to use COQPATH is if you have two separate Coq projects which you don't want to install, but you would like to use one from the other as if it has been installed. However, I still think explaining this to a starting user is pointless, and it should be a technical packaging detail.

Reply:
Let's go through this a bit:

> If you have a coq_makefile project or dune project, they will both install into user-contrib today by default.

That was not what I noticed when I tried this two months ago, though I was working from my own build out of the source tree.  At the moment I find `coq_makefile` won't even run for me with a fresh 8.15.2 installed with opam.  See #16183.  What is your command line to make for the coq_makefile-generated makefile?

(How exactly can I build and test coq_makefile locally when there's no dune target for it?  I need to do something like a `dune install`, then run the installed version.)

IIRC, building with coq_makefile left the .vo files in the same directory as the .v file.  You certainly don't want to put the source code you're busily modifying into the Coq installation directory, let alone its git files.

On general principles, I think it's a poor practice to copy package(s) you're in the midst of modifying into the official installation directories--it's too easy to mess something up even if its scripted.  The installation directories should changed only by opam or when you explicitly install a package.

For example, suppose you use bignums for your work and you also happen to be a maintainer of bignums.  It will be all too easy to end up with the wrong binaries for bignums in the installation directory.  If you upgrade Coq with Coq platform, the modified files will, in effect, disappear because it will be in a new switch.  Not sure what would happen with opam if you change the coq version or a package version directly.

The IDEs will find the _CoqProject associated with a script file based on the file path, but that won't work if a) you try to access the package from an unnamed buffer or b) from a file that's outside of your package.

Let me know your thoughts.  There are some facts to check, maybe you can see if `coq_makefile` and the file it generates works for you.  Two plus months after I wrote this, it takes a while to reconstruct my reasoning.

> The `Cd` commands in Coq are really ready to be deprecated any day. 

Yeah, I saw that.  It may be useful as IDE-only concept rather than as a command.  The doc says "The `<>` in the final entry represents an empty logical pathname, which permits loading files from the associated directory with just the basename of the script file, e.g. specify `Foo` to load `Foo.vo`.  Note that the :cmd:`Cd` command doesn't change the associated directory--you would need to restart CoqIDE."

Reply:
> Let's go through this a bit:
> 
> > If you have a coq_makefile project or dune project, they will both install into user-contrib today by default.
> 
> That was not what I noticed when I tried this two months ago, though I was working from my own build out of the source tree. At the moment I find `coq_makefile` won't even run for me with a fresh 8.15.2 installed with opam. See #16183. What is your command line to `make` for the coq_makefile-generated makefile?

If I generate a makefile with `coq_makefile -f _CoqProject -o Makefile` I will get a makefile called `Makefile`. If I then run `make` in the same directory then the project will be built. To install it, the makefile will contain an `install` target which I can run using `make install`. This will the proceed to install in the install directory.

If you installed Coq via opam, then you will have likely come across a command such as `eval $(opam env)`. This is very important. If you run `opam env` you will see it output some text about setting some environment variables. The `eval $(..)` part will set those variables for your terminal session. If you set up opam (during opam init) to configure something in your user profile, then this will automatically be run on login.

The important environment variable that gets set here is OCAMLPATH. This is the install location for OCaml libraries. Coq is an OCaml library so it can be found in `$OCAMLPATH/[lib/?]Coq`. (I'm not sure about the lib part). Inside the Coq directory there is the user-contrib directory in which Coq libraries themselves get installed.

So how does Coq know about the install directory (in coq_makefile)? It first checks OCAMLPATH for the Coq folder. After which it has access to theories/ and user-contrib/. It then installs into the second.

The Coq folder in OCAMLPATH can be overridden by specifying COQLIB. Which for the most part is unset. If it is set Coq will take precedence for that over the guessing game above.

> (How exactly can I build and test coq_makefile locally when there's no dune target for it? Do I need to do something like a `dune install`, then run the installed version?)

I think there might be some confusion between the actual Coq targets, and the shims. Let me try to clear up some of this.

As I described above, Coq when starting, does a lot of guesswork trying to understand where the Coq stdlib might be and where the installation locations are. Part of its validation for the guess of COQLIB is to look for a very specific file in the stdlib called `theories/Init/Prelude.vo`.

This is the "Coq prelude" which is implicitly (`Require`d) in every Coq session. Therefore Coq needs to know where it is in order to use it. You can disable this implicit loading with the `-noinit` option.

So why do we need shims? Well if you load a Coq binary, without the prelude file present, you will get a message saying something like "coqlib not found". This is kind of strange, since coq_makefile doesn't even do any proof assisting, it is just a tool to generate a makefile. The answer to why this is, is that most of the Coq binaries share the same `boot/env.ml` file which takes care of setting up the COQLIB business I described above.

This is not an ideal factorization, since it means that coq_makefile has to do some unrelated things. But it historically allowed for sharing functions such as printing command line help options etc.

Back to the shims: The `dev/shim/` folder's job is to make sure enough of the Coq prelude is built, so that the binaries do not complain. That is all that it does. To get the same effect as say `dune build dev/shim/coqc-prelude` you need to build both the prelude and the `coqc` binary. This can be done first by writing `dune build theories/Init/Prelude.vo`, Dune will then go and starting building all the necessary things and eventually that.

Next you want to build the `coqc` binary. If you try `dune build coqc` you will see that it does not work. The reason is that there is no build target named `coqc` in the Coq directory. What there is instead is an executable called `coqc` which will live somewhere in the Coq source tree. It is not important to locate these executables, Dune can do it for you. This is what the `dune exec -- coqc` command is doing. The `--` just means "parse the rest as a single command line arg", this makes passing args to coqc easy.

`dune exec` first goes and looks for an executable called `coqc`, which is declared in https://github.com/coq/coq/blob/master/topbin/dune. It then tries to build it and run it.

All of this is exactly what the `dev/shim/coqc-prelude` shim is doing, but with some environment state also being set.

```lisp
(rule
 (targets coqc-prelude)
 (deps
  %{bin:coqc}
  %{project_root}/theories/Init/Prelude.vo)
 (action
  (with-stdout-to coqc-prelude
   (progn
    (echo "#!/usr/bin/env bash\n")
    (bash "echo '\"$(dirname \"$0\")\"/%{bin:coqc} -I \"$(dirname \"$0\")/%{project_root}/../install/default/lib\" -coqlib \"$(dirname \"$0\")\"/%{project_root} \"$@\"'")
    (run chmod +x %{targets})))))
```

You can see that the `(rule )` stanza declares `%{bin:coqc}` (the coqc binary), and `Prelude.vo` as dependencies. This means these need to get built before you do what is described in the `(action )` field. All that is happening in the action field however is `coqc` is being run, with an explicit `-coqlib` and (`-I` for coq plugins) being set to the ones that Dune just built. This allows you to run your executable in the correct env.

Very confusingly, it is actually creating a wrapper bash script around coqc with the env variables in that. (The with-stdout-to part pipes the output (of echo) to a file named `coqc-prelude`. The `(progn )` part runs several command in succession. You can even see the `chmod` part letting you run the script at the end.

Let's come back to coq_makefile. How can we create a shim for that? Well it's probably very easy, just replace mention of coqc in the above thing with `coq_makefile`. You also probably don't need the `-I` part but I don't think it does any harm. `coq_makefile` also requires some extra files such as `tools/Makefile.in` IIRC, so you need to add that to the deps field. In Dune a good rule of thumb is to always declare your deps, even if you know they will always be there.

> IIRC, building with coq_makefile left the .vo files in the same directory as the .v file. You certainly don't want to put the source code you're busily modifying into the Coq installation directory, let alone its git files.

This is how makefile builds work unfortunately. They will produce the .vo targets in the same directory as the .v files. However, the installation part is exactly the job of `make install` which will systematically copy and install the `.vo` files to the installation directory it determines in the mess above.

> On general principles, I think it's a poor practice to copy package(s) you're in the midst of modifying into the official installation directories--it's too easy to mess something up even if its scripted. The installation directories should changed only by opam or when you explicitly install a package.

Absolutely, nothing should have to be installed when developing a package.

> For example, suppose you use bignums for your work and you also happen to be a maintainer of bignums. It will be all too easy to end up with the wrong binaries for bignums in the installation directory. If you upgrade Coq with Coq platform, the modified files will, in effect, disappear because it will be in a new switch. Not sure what would happen with opam if you change the coq version or a package version directly.

Typically, if you are maintaining the packages in Coq platform, then you probably should not be using Coq platform to do this. Coq platform is supposed to be an easy-to-install multi-platform solution for getting Coq working, rather than a development environment.

Something that might be of interest is the recent inter-project composition for Coq feature that Rudi and I worked on in Dune. This let's you say work on bugnums and another project that depends on it simultaneously, since Dune now knows how to "compose" and build these projects in tandem. Getting smooth IDE support for this is something that needs to be worked on.

> The IDEs will find the _CoqProject associated with a script file based on the file path, but that won't work if a) you try to access the package from an unnamed buffer or b) from a file that's outside of your package.

I can't speak for PG, but coqidetop based IDEs (CoqIDE and VSCoq) have some trouble with _CoqProject files. It only really works when you open a named buffer in the scope of a _CoqProject file. Even if you open a new buffer and save it, I find myself having to reopen before the _CoqProject is properly detected.

For multi-project developments, that is all too true. CoqIDE really doesn't work that well. One solution is to have two instances of CoqIDE running, per _CoqProject file. This can typically have better results.

> Let me know your thoughts. There are some facts to check, maybe you can see if `coq_makefile` and the file it generates works for you. Two plus months after I wrote this, it takes a while to reconstruct my reasoning.

I've posted a lot of information in this reply, so hopefully it is useful. I think if we iterate on this, we will have much better documentation than before.

> > The `Cd` commands in Coq are really ready to be deprecated any day.
> 
> Yeah, I saw that. It may be useful as IDE-only concept rather than as a command. The doc says "The `<>` in the final entry represents an empty logical pathname, which permits loading files from the associated directory with just the basename of the script file, e.g. specify `Foo` to load `Foo.vo`. Note that the :cmd:`Cd` command doesn't change the associated directory--you would need to restart CoqIDE."

The `Cd` command is really an archaic artefact from the time people used `coqtop` to write Coq files. It is terrible for `coqc` for example. It actually changes where the `.vo` file will get ouput!

Reply:
Thanks for the detailed response.  Perhaps a bit too much detail.  I hoped that one of you experts would eventually create a `coq_makefile` shim to bring dune closer to parity with `make`-based builds.

The main point is that given the behavior of the makefile generated by `coq_makefile`, users need to know about COQPATH.  Agreed?

> Typically, if you are maintaining the packages in Coq platform, then you probably should not be using Coq platform to do this. Coq platform is supposed to be an easy-to-install multi-platform solution for getting Coq working, rather than a development environment.

Huh?  If your package only has .v files and no OCaml, you don't need to have the Coq source tree.  

> Something that might be of interest is the recent inter-project composition for Coq feature that Rudi and I worked on in Dune.

An interesting use case, but let's discuss after this PR is completed.  May need some tweaks to Coq.

> coqidetop based IDEs (CoqIDE and VSCoq) have some trouble with _CoqProject files. It only really works when you open a named buffer in the scope of a _CoqProject file. Even if you open a new buffer and save it, I find myself having to reopen before the _CoqProject is properly detected.

The _CoqProject parameters become CLI arguments to coqtopide.  The only way to change them is to start a new coqtopide.  Resetting the buffer is sufficient and less annoying than closing and reopening.  This can be improved, CoqIDE should be checking for modified files.  Also you ought to be able to edit _CoqProject inside CoqIDE in a buffer.

> For multi-project developments, ... CoqIDE really doesn't work that well. One solution is to have two instances of CoqIDE running, per _CoqProject file. This can typically have better results.

If you run two instances, changes made in instance 1 are not visible in instance 2 until you reset the buffer and create a new coqtopide.  This can also be improved.

Reply:
> COQPATH is not really needed by anybody.

COQPATH is absolutely essential in Fiat Crypto and it's dependencies, which can be configured to use either installed dependencies or to use submodules for dependencies.  We set COQPATH to make use of submodule dependencies, which are often built with different flags than the main project.

Reply:
Agreed, COQPATH is needed.  I recommend its use in the utilities chapter.  @Alizter, if you still disagree, please speak up.

Reply:
What I meant is that it is not essential to use COQPATH in order to load installed libraries. I did not intend to sound like it was unused.